dll/win32/shell32/dialogs/filetypes.cpp

```diff
@@ -23,27 +23,150 @@
-//        Cmd: rundll32.exe shell32.dll,Options_RunDLL 0
+// rundll32.exe shell32.dll,Options_RunDLL 0
 
 /////////////////////////////////////////////////////////////////////////////
 
+#ifndef FTA_OpenIsSafe
+#define FTA_NoEdit       0x00000008
+#define FTA_NoRemove     0x00000010
+#define FTA_NoNewVerb    0x00000020
+#define FTA_NoEditVerb   0x00000040
+#define FTA_NoRemoveVerb 0x00000080
+#define FTA_NoEditDesc   0x00000100
+#define FTA_NoEditIcon   0x00000200
+#define FTA_NoEditDflt   0x00000400
+#define FTA_OpenIsSafe   0x00010000
+#endif
+#define FTA_MODIFYMASK (FTA_OpenIsSafe) // Bits touched by EditTypeDlg
+
+#define NOASSOCRESID IDI_SHELL_DOCUMENT
+#define SUPPORT_EXTENSIONWITHOUTPROGID 1 // NT5 does not support these but NT6 does
+
+#define ASSOC_CCHMAX (32 + 1) // Extension or protocol (INTERNET_MAX_SCHEME_LENGTH)
+#define TYPENAME_CCHMAX max(100, RTL_FIELD_SIZE(SHFILEINFOA, szTypeName))
+#define VERBKEY_CCHMAX (127 + 1)
+#define ICONLOCATION_CCHMAX (MAX_PATH + 1 + 11)
+
 typedef struct FILE_TYPE_ENTRY
 {
-    WCHAR FileExtension[30];
-    WCHAR FileDescription[100];
+    WCHAR FileExtension[ASSOC_CCHMAX];
+    WCHAR FileDescription[TYPENAME_CCHMAX];
     WCHAR ClassKey[MAX_PATH];
-    WCHAR ClassName[64];
     DWORD EditFlags;
     WCHAR AppName[64];
-    HICON hIconLarge;
     HICON hIconSmall;
     WCHAR ProgramPath[MAX_PATH];
     WCHAR IconPath[MAX_PATH];
     INT nIconIndex;
+
+    bool IsExtension() const
+    {
+        return FileExtension[0] == '.';
+    }
+    LPCWSTR GetAssocForDisplay() const
+    {
+        return FileExtension + (FileExtension[0] == '.');
+    }
+    void InvalidateTypeName()
+    {
+        FileDescription[0] = FileDescription[1] = UNICODE_NULL;
+    }
+    void InvalidateDefaultApp()
+    {
+        ProgramPath[0] = ProgramPath[1] = AppName[0] = AppName[1] = UNICODE_NULL;
+    }
+    void Initialize()
+    {
+        ClassKey[0] = UNICODE_NULL;
+        IconPath[0] = UNICODE_NULL;
+        nIconIndex = 0;
+        InvalidateTypeName();
+        InvalidateDefaultApp();
+    }
+    void DestroyIcons()
+    {
+        if (hIconSmall)
+            DestroyIcon(hIconSmall);
+        hIconSmall = NULL;
+    }
 } FILE_TYPE_ENTRY, *PFILE_TYPE_ENTRY;
 
+typedef struct {
+    HIMAGELIST himlSmall;
+    UINT IconSize;
+    HICON hDefExtIconSmall;
+    HBITMAP hOpenWithImage;
+    WCHAR DefExtTypeNameFmt[TYPENAME_CCHMAX];
+    WCHAR NoneString[33];
+    INT8 SortCol, SortRev;
+    UINT Restricted;
+} FILE_TYPE_GLOBALS, *PFILE_TYPE_GLOBALS;
+
+static BOOL
+RegValExists(HKEY hKey, LPCWSTR Name)
+{
+    return !RegQueryValueExW(hKey, Name, 0, NULL, NULL, NULL);
+}
+
+static BOOL
+RegKeyExists(HKEY hKey, LPCWSTR Path)
+{
+    return !RegOpenKeyExW(hKey, Path, 0, MAXIMUM_ALLOWED, &hKey) && !RegCloseKey(hKey);
+}
+
+static HRESULT
+SHELL32_GetRegDW(HKEY hKey, LPCWSTR Name, DWORD &Val, DWORD Def = 0, BOOL Strict = FALSE)
+{
+    DWORD rt, cb = 4, ec = RegQueryValueExW(hKey, Name, 0, &rt, (BYTE*)&Val, &cb);
```

Comment:
`sizeof(rt)` is the wrong "source" but I get your point.

From my point of view, the desired value is 4 and it will always be 4 until the end of time...